### ISSUE TITLE
Narrow array element

### ISSUE THREAD

**Issue by phanen** [EmmyLuaLs/emmylua-analyzer-rust#886]
```lua
---@type string[]
local a = {}

-- if #a == 0 then return end
if not a[1] then return end

-- ---@type string
-- local s = a[1]

---@type string
local s = a[#a]
```

* **Expected**: no warning
* **Actual**: `Cannot assign string? to string.`
